include/mbedtls/aes.h

```diff
@@ -69,26 +69,50 @@ extern "C" {
 // Regular implementation
 //
 
+/** VIA padlock needs 16 extra bytes for alignment. */
+#if (defined(MBEDTLS_PADLOCK_C) && defined(MBEDTLS_HAVE_X86)) ||        \
+    (defined(MBEDTLS_AESNI_C) && MBEDTLS_AESNI_HAVE_CODE == 2)
```

Comment:
Note that there's a reason why `MBEDTLS_AESNI_HAVE_CODE` is defined in a private header: it depends on the compiler/options that the library was built with. The application may be compiled with a different compiler. `MBEDTLS_AESNI_HAVE_CODE` must not affect the ABI.

It's ok for `MBEDTLS_AESNI_C` and `MBEDTLS_PADLOCK_C` to affect the ABI because we made them part of the API (even though the choice of method for calculating AES is arguably something that could have stayed hidden, so that you deploy the same application binary and link it against different library binaries on different hardware variants).

It's not ok for `MBEDTLS_AESNI_HAVE_CODE` to affect the ABI because the exact same library configuration can end up being used with either of the two AESNI implementations. `MBEDTLS_AESNI_HAVE_CODE` is defined in a private header for that reason.

If `aes.c`'s `MAY_NEED_TO_ALIGN` becomes public, it can depend on `MBEDTLS_AESNI_C` but not on the choice of AESNI implemetation. Even so, this may make some people unhappy, because until now AESNI and AESCE support didn't affect the ABI. However, linking an application compiled with a different configuration is not a thing we officially support, so those people were doing something wrong.